internal/filter/filter.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package filter defines which tickets pass which filters.  Other implementations which help
+// filter tickets (eg, a range index lookup) must conform to the same set of tickets being within
+// the filter as here.
+package filter
+
+import (
+	structpb "github.com/golang/protobuf/ptypes/struct"
+	"open-match.dev/open-match/internal/pb"
+)
+
+// Filter returns all of the passed tickets which are included by all the passed filters.
+func Filter(tickets []*pb.Ticket, filters []*pb.Filter) []*pb.Ticket {
+	var result []*pb.Ticket
+
+	for _, t := range tickets {
+		if InFilters(t, filters) {
+			result = append(result, t)
+		}
+	}
+
+	return result
+}
+
+// InFilters returns if the given ticket is included in all of the passed filters.
+func InFilters(ticket *pb.Ticket, filters []*pb.Filter) bool {
+	for _, f := range filters {
+		if !InFilter(ticket, f) {
+			return false
+		}
+	}
+	return true
+}
+
+// InFilter returns if the given ticket is included in the given filter.
+func InFilter(ticket *pb.Ticket, filter *pb.Filter) bool {
+	if ticket.Properties == nil || ticket.Properties.Fields == nil {
```

Comment:
I think we can move the sanity checks to our `mmlogic_service.go` file. It is the caller's responsibility to make sure the inputs are valid.

Reply:
I want this here for a few reasons:

- Not having this check increases the complexity of cross package contracts.
- This package is the definition of how tickets are filtered.  This statement specifies how improper tickets are handled.
- It's actually valid to not have any properties.  If you ask for a tickets with no filters, tickets with no properties are returned.
- The call code will mostly be calling Filter, which contains a list of tickets.  This means callers don't have to do pre-filtering.

Reply:
1. The reason that I wan to move this check is similar sanity checks are done at service level (e.g. [1](https://github.com/googleforgames/open-match/blob/master/internal/app/backend/backend_service.go#L61), [2](https://github.com/googleforgames/open-match/blob/master/internal/app/frontend/frontend_service.go#L49), [3](https://github.com/googleforgames/open-match/blob/master/internal/app/mmlogic/mmlogic_service.go#L53)). They guarantee even if a user pass in an invalid request our service won't panic.
2. I think it is fair to define this package as "how valid tickets are filtered".
3. It is valid to have an empty property bag as long as ticket.Properties != nil (which is checked in frontend_service.go), therefore any call codes with tickets from the storage service already satisfied this sanity check.